src/roslyn/src/Compilers/Core/MSBuildTask/ManagedToolTask.cs

```diff
@@ -239,10 +240,18 @@ internal static string GetBuildTaskDirectory()
         protected override bool ValidateParameters()
         {
             // Set DOTNET_ROOT so that the apphost executables launch properly.
+            // Unset all other DOTNET_ROOT* variables so for example DOTNET_ROOT_X64 does not override ours.
             if (RuntimeHostInfo.GetToolDotNetRoot() is { } dotNetRoot)
             {
                 Log.LogMessage("Setting {0} to '{1}'", RuntimeHostInfo.DotNetRootEnvironmentName, dotNetRoot);
-                EnvironmentVariables = [.. EnvironmentVariables ?? [], $"{RuntimeHostInfo.DotNetRootEnvironmentName}={dotNetRoot}"];
+                EnvironmentVariables =
+                [
+                    .. EnvironmentVariables?.Where(static e => !e.StartsWith(RuntimeHostInfo.DotNetRootEnvironmentName, StringComparison.OrdinalIgnoreCase)) ?? [],
+                    .. Environment.GetEnvironmentVariables().Cast<System.Collections.DictionaryEntry>()
+                        .Where(e => ((string)e.Key).StartsWith(RuntimeHostInfo.DotNetRootEnvironmentName, StringComparison.OrdinalIgnoreCase))
+                        .Select(e => $"{e.Key}="),
```

Comment:
Could `DOTNET_ROOT` be in this list already? Is it ok to set both
```
DOTNET_ROOT=
DOTNET_ROOT={dotNetRoot}
```
Does last one win?

Reply:
Yes, it could be in the list but the last one wins